LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ConcentratorDeduplication.cs

```diff
@@ -30,96 +36,103 @@ internal sealed class ConcentratorDeduplication<T> :
 
         public ConcentratorDeduplication(
             IMemoryCache cache,
-            WebSocketWriterRegistry<StationEui, string> socketRegistry,
-            ILogger<IConcentratorDeduplication<T>> logger)
+            IDeduplicationStrategyFactory deduplicationStrategy,
+            ILogger<IConcentratorDeduplication> logger)
         {
             this.cache = cache;
-            this.socketRegistry = socketRegistry;
+            this.deduplicationStrategy = deduplicationStrategy;
             this.logger = logger;
         }
 
-        public bool ShouldDrop(T frame, StationEui stationEui)
+        public Result CheckDuplicate(LoRaRequest loRaRequest, LoRaDevice? loRaDevice)
         {
-            if (frame == null) throw new ArgumentNullException(nameof(frame));
+            _ = loRaRequest ?? throw new ArgumentNullException(nameof(loRaRequest));
 
-            var key = CreateCacheKey(frame);
+            var key = CreateCacheKey(loRaRequest);
+            var stationEui = loRaRequest.StationEui;
 
             StationEui previousStation;
             lock (this.cache)
```

Comment:
locking on the cache means we don't control the lock. Anyone, injecting the memory cache instance, could be locking on it. I would suggest adding a dedicated lock on a private field we control. Having a public accessible object we lock on can have very hard to track bugs like deadlocks (when nested locking occurs for example).

Reply:
Right, as already discussed 🤦‍♂️ Changed in 0a8c12d482b230cb20195f4152e3eefc59b8c318 to use a `SemaphoreSlim`